sdks/python/container/boot.go

```diff
@@ -105,18 +107,57 @@ func main() {
 
 	// (2) Retrieve and install the staged packages.
 
-	dir := filepath.Join(*semiPersistDir, "staged")
+	func() {
 
-	files, err := artifact.Materialize(ctx, *artifactEndpoint, info.GetRetrievalToken(), dir)
-	if err != nil {
-		log.Fatalf("Failed to retrieve staged files: %v", err)
-	}
+        installCompleteFile := filepath.Join(os.TempDir(), "beam.install.complete")
 
-	// TODO(herohde): the packages to install should be specified explicitly. It
-	// would also be possible to install the SDK in the Dockerfile.
-	if setupErr := installSetupPackages(files, dir); setupErr != nil {
-		log.Fatalf("Failed to install required packages: %v", setupErr)
-	}
+		// skip if install already complete
+		_, err = os.Stat(installCompleteFile)
+		if err == nil {
+			return
+		}
+
+		// lock to guard from concurrent artifact retrieval and installation,
+		// when called by child processes in a worker pool
+		lock, err := lockfile.New(filepath.Join(os.TempDir(), "beam.install.lck"))
+		if err != nil {
+			log.Fatalf("Cannot init artifact retrieval lock: %v", err)
+		}
+
+		for err = lock.TryLock(); err != nil; err = lock.TryLock() {
+			switch err {
+			case lockfile.ErrBusy, lockfile.ErrNotExist:
```

Comment:
Nit: This is fine as is, but here's an optional suggestion.
In this instance the lockfile package author has a type TemporaryError to indicate the error can be retried,which we can check for instead of listing all temporary errors.

if _, ok := err.(lockfile.TemporaryError); ok {
  //sleep etc
} else {
  log.Fatalf(...
}

Or if you're keen on a switch, how about a type switch?
switch err.(type) {
case lockfile.TemporaryError:
  //sleep etc
default:
  log.Fatalf(...
}